Add executable for CurvedScalarWave

prayush opened the pull request:
## Proposed changes

- Adding executables to evolve the curved scalar wave system
- Periodic boundary conditions imposed on the outer boundaries, so will work only with periodic domains for now.

~~**Note**: The first commit makes up #2972 . Please review from 2nd commit onward.~~

~~**Note**: The first 3 commits make up #3100 and #3116 . Please review commits after those.~~

~~**Note**: The first commit make up #3100 . Please review commits after those.~~

**Note**: All dependencies of this PR are on develop now.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

nikwit commented on src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/PlaneWaveMinkowski.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "PointwiseFunctions/AnalyticData/CurvedWaveEquation/PlaneWaveMinkowski.hpp"
+
+/// \cond
+namespace CurvedScalarWave ::AnalyticData {
```
(optional) can you run clang-format over this?

prayush replied:
done

nikwit commented on tests/InputFiles/CurvedScalarWave/PlaneWaveMinkowski1D.yaml:
```diff
@@ -0,0 +1,76 @@
+          # Based on the step size choices above, this should aim for
+          # about 100 steps per slab.
+          - Cfl:
+              SafetyFactor: 20
```
have you considered adding some actual observations for testing purposes?

prayush replied:
added to the 2D case

nikwit commented on tests/InputFiles/CurvedScalarWave/PlaneWaveMinkowski2D.yaml:
```diff
@@ -0,0 +1,66 @@
+DomainCreator:
+  Rectangle:
+    LowerBound: [0.0, 0.0]
+    UpperBound: [16.283185307179586, 16.283185307179586]
```
should this be 2 pi as well?

prayush replied:
yes, changed

nikwit commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -0,0 +1,355 @@
+      "initial_data must be either an analytic_data or an analytic_solution");
+
+  using system = CurvedScalarWave::System<Dim>;
+  static constexpr dg::Formulation dg_formulation =
+      dg::Formulation::StrongInertial;
+  using temporal_id = Tags::TimeStepId;
```
you can probably delete these two lines

prayush replied:
done

nikwit commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -0,0 +1,355 @@
+                  Phase, Phase::Evolve,
+                  tmpl::list<
+                      Actions::RunEventsAndTriggers, Actions::ChangeSlabSize,
+                      tmpl::conditional_t<
```
remove this to fix the segfault, see #2862

prayush replied:
done

nikwit commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -0,0 +1,355 @@
+                                     Actions::MutateApply<boundary_scheme>>,
+                          tmpl::list<Actions::MutateApply<boundary_scheme>,
+                                     Actions::RecordTimeStepperData<>>>,
+      Actions::UpdateU<>,
```
move this into the local_time_stepping == false branch to fix the segfault, see #2862

prayush replied:
done

nikwit commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -0,0 +1,355 @@
+      Initialization::Actions::TimeStepperHistory<EvolutionMetavars>,
+      CurvedScalarWave::Actions::InitializeConstraintDampingGammas,
+      CurvedScalarWave::Actions::InitializeGrVars<volume_dim>,
+      tmpl::list<dg::Actions::InitializeInterfaces<
```
remove this `tmpl::list<>`

prayush replied:
done

nikwit commented on src/Evolution/Systems/CurvedScalarWave/Constraints.hpp:
```diff
@@ -13,10 +13,46 @@
+ * Can be retrieved using `CurvedScalarWave::Tags::ConstraintGamma1`
+ * and `CurvedScalarWave::Tags::ConstraintGamma2`.
+ */
+struct ConstraintGamma1Compute : ConstraintGamma1, db::ComputeTag {
```
Do these need to be ComputeTags? can you not do this the same way as in `ScalarWave`, where it is in `Initialize.hpp`?

prayush replied:
made them simple tags that are initialized in `Initialize.hpp`

nikwit commented on src/Evolution/Systems/CurvedScalarWave/Initialize.hpp:
```diff
@@ -0,0 +1,183 @@
+/// `Initialization` phase action list prior to this action.
+template <size_t Dim>
+struct InitializeGrVars {
+  using compute_tags =
```
can these not all be `SimpleTag`s as well since we are evolving on a fixed background?

prayush replied:
yup, fixed

nikwit commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -0,0 +1,355 @@
+// Distributed under the MIT License.
```
would it be possible to combine the Metavariables of the CurvedScalarWave with the ScalarWave Metavariables? They seem extremely similar except the interface initialization.

prayush replied:
Instead of merging them, we can actually just remove the `ScalarWave` system completely and just use the `CurvedScalarWave` system through and through with a `Minkowski` background. The reason we keep the `ScalarWave` system at all is to treat it as an illustrative / educational tool. This was discussed this on the spectre telecon some time ago, and we decided to leave the `CurvedScalarWave` system independent of the `ScalarWave` executable, and neither merge the two nor get rid of the flat space system for the time being

nikwit reviewed (changes requested):
group code review with @nilsleiffischer. looks good except some minor changes

nilsdeppe reviewed (commented):
I think it would be good to update the CurvedSW system to use the updated evolution scheme. I think this will mostly require a refactor of the penalty boundary correction (see e.g. scalar wave or GH for an example), and for initial boundary conditions it would be dirichlet analytic and periodic that probably make the most sense (see scalar wave and Gh for examples)

Edit: I'm happy to have a call to discuss the changes if that would be helpful :)

nikwit commented:
@prayush @nilsdeppe I am just looking into adding the new boundary conditions/corrections scheme to the executable in this form as I need it for another project (including the constraint-preserving boundary conditions). If Prayush would like for me to take this off his hands I would be happy to do it - let's just avoid us both doing the work :) Im sure a call would be helpful in this case, Nils.

kidder commented:
@prayush @nikwit @nilsdeppe what is this PR waiting on?

prayush committed:
Add executables for the CurvedScalarWave system

 - Exe: EvolveCurvedPlaneWaveMinkowski{1,2,3}D
 - Include input files
 - Support for periodic BCs (only) on external boundaries

prayush commented:
> @prayush @nikwit @nilsdeppe what is this PR waiting on?

This is now waiting on #3100 ~~and #3116~~ and has been updated to use the updated evolution scheme.

prayush committed:
fixup: add GR quantities as simple tags instead of compute tags

prayush committed:
fixup: rebase onto develop

prayush committed:
fixup: remove time stepper tag from global cache

nikwit commented:
@prayush this needs a rebase now

prayush committed:
fixups from rebase

prayush commented:
> > @prayush @nikwit @nilsdeppe what is this PR waiting on?
> 
> This is now waiting on ~#3100 and #3116~ and has been updated to use the updated evolution scheme.

ping!

nikwit commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -0,0 +1,354 @@
+                 observers::ObserverWriter<EvolutionMetavars>,
+                 dg_element_array>;
+
+  static constexpr Options::String help{
```
can you update this? "Evolve a scalar wave in Dim spatial dimension on a curved background." You should not need to mention the flux here as it is now a runtime choice.

nikwit commented on src/Evolution/Executables/CurvedScalarWave/EvolveCurvedScalarWave.hpp:
```diff
@@ -0,0 +1,354 @@
+    &domain::creators::register_derived_with_charm,
+    &domain::creators::time_dependence::register_derived_with_charm,
+    &domain::FunctionsOfTime::register_derived_with_charm,
+    &CurvedScalarWave::BoundaryCorrections::register_derived_with_charm,
```
this should be updated to the new registration mechanism I think. I would be fine with doing it in a follow-up PR.

nikwit commented on src/Evolution/Systems/CurvedScalarWave/BoundaryConditions/RegisterDerived.cpp:
```diff
@@ -0,0 +1,15 @@
+// Distributed under the MIT License.
```
If I understand the new registration mechanism correctly these files are no longer needed.

nikwit commented on src/Evolution/Systems/CurvedScalarWave/Initialize.hpp:
```diff
@@ -0,0 +1,211 @@
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const auto& mesh = db::get<domain::Tags::Mesh<Dim>>(box);
+    Scalar<DataVector> gamma1{mesh.number_of_grid_points(), 0.};
```
do you think we should make these values a runtime choice? It seems odd to hard-code them.

nikwit commented on tests/InputFiles/CurvedScalarWave/PlaneWaveMinkowski3D.yaml:
```diff
@@ -0,0 +1,69 @@
+# Distributed under the MIT License.
```
I think I will update the 3D version to do a proper scalar wave simulation in Kerr Schild background including proper boundary conditions and some interpolation observations. For now I am fine with adding a simple version on a periodic domain like here.

nilsdeppe commented on src/Evolution/Systems/CurvedScalarWave/Initialize.hpp:
```diff
@@ -0,0 +1,211 @@
+
+namespace CurvedScalarWave {
+namespace Actions {
+/// \ingroup InitializationGroup
```
I'm not seeing tests for these actions. Is that just me overlooking them?

nilsdeppe commented:
Replaced by #3729
